src/background/tab-context-manager.ts

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { BrowserAdapter } from '../common/browser-adapters/browser-adapter';
+import { Message } from '../common/message';
+import { BrowserMessageBroadcasterFactory } from './browser-message-broadcaster-factory';
+import { ExtensionDetailsViewController } from './extension-details-view-controller';
+import { TabToContextMap } from './tab-context';
+import { TabContextFactory } from './tab-context-factory';
+
+export class TabContextManager {
+    constructor(
+        private readonly targetPageTabIdToContextMap: TabToContextMap,
+        private readonly broadcasterFactory: BrowserMessageBroadcasterFactory,
+        private readonly browserAdapter: BrowserAdapter,
+        private readonly detailsViewController: ExtensionDetailsViewController,
+        private readonly tabContextFactory: TabContextFactory,
+        private persistStoreData = false,
+    ) {}
+
+    public addTabContextIfNotExists(tabId: number): void {
+        if (this.targetPageTabIdToContextMap[tabId] === undefined) {
+            this.targetPageTabIdToContextMap[tabId] = this.tabContextFactory.createTabContext(
+                this.broadcasterFactory.createTabSpecificBroadcaster(tabId),
+                this.browserAdapter,
+                this.detailsViewController,
+                tabId,
+                this.persistStoreData,
+            );
+        }
+    }
+
+    public async deleteTabContext(tabId: number): Promise<void> {
+        const tabContext = this.targetPageTabIdToContextMap[tabId];
+        if (tabContext) {
+            delete this.targetPageTabIdToContextMap[tabId];
```

Comment:
nit: can we ever get into a situation where we have a tab ID key mapping to null/ undefined? If so, this would leave the key in the map instead of removing it, is that what we want? Same with in `addTabContextIfNotExists`, the if check doesn't check if the key exists or not, which seems to be a subtle change in behavior.

Reply:
Good catch, I didn't realize those were different. I can't think of a situation where we would expect an undefined TabContext in this map, but I'll change it for consistency with what we had before.